controllers/AdminController.php

```diff
@@ -294,23 +309,40 @@ public function actionInfo($id)
 
     /**
      * Switches to the given user for the rest of the Session.
+     * When no id is given, we switch back to the original admin user
+     * that started the impersonation.
      *
      * @param int $id
      *
      * @return string
      */
-    public function actionSwitch($id)
+    public function actionSwitch($id = null)
     {
-        $user = $this->findModel($id);
+        if(!Yii::$app->getModule('user')->enableImpersonateUser)
+            throw new ForbiddenHttpException(Yii::t('user', 'Impersonate user is disabled in the application configuration'));
+
+        $key = self::ORIGINAL_USER_SESSION_KEY;
 
-        $old = \Yii::$app->user;
+        if(!$id && Yii::$app->session->has($key)) {
+            $user = $this->findModel(Yii::$app->session->get($key));
+            Yii::$app->session->remove($key);
+        } else {
+            if (!Yii::$app->user->identity->isAdmin)
+                throw new ForbiddenHttpException;
+
+            $user = $this->findModel($id);
+            Yii::$app->session->set($key, Yii::$app->user->id);
+        }
 
-        \Yii::$app->user->switchIdentity($user, 3600);
+        Yii::$app->user->switchIdentity($user, 3600);
 
-        \Yii::warning(sprintf('User %s(id: %d) switched to user %s(id: %d).',
-                $old->identity->username, $old->id, \Yii::$app->user->identity->username, \Yii::$app->user->id));
+        Yii::warning(sprintf('User %s(id: %d) switched to user %s(id: %d).',
```

Comment:
I added events (https://github.com/dektrium/yii2-user/pull/820/commits/c74b26ea5b93e990561f5a0bb0edaaa1128aaee8) for devs to implement logging by themselves, if they want to.